easybuild/easyblocks/c/cudacompat.py

```diff
@@ -0,0 +1,143 @@
+        )
+        super(EB_CUDAcompat, self).fetch_step(*args, **kwargs)
+
+    def extract_step(self):
+        """Extract the files without running the installer."""
+        execpath = self.src[0]['path']
+        tmpdir = os.path.join(self.builddir, 'tmp')
+        targetdir = os.path.join(self.builddir, 'extracted')
+        run_cmd("/bin/sh " + execpath + " --extract-only --tmpdir='%s' --target '%s'" % (tmpdir, targetdir))
+        self.src[0]['finalpath'] = targetdir
+
+    def install_step(self):
+        """Install CUDA compat libraries by copying library files and creating the symlinks."""
+        libdir = os.path.join(self.installdir, 'lib')
+        mkdir(libdir)
+
+        # From https://docs.nvidia.com/deploy/cuda-compatibility/index.html#installing-from-network-repo:
+        # The cuda-compat package consists of the following files:
+        #   - libcuda.so.* - the CUDA Driver
+        #   - libnvidia-nvvm.so.* - JIT LTO ( CUDA 11.5 and later only)
+        #   - libnvidia-ptxjitcompiler.so.* - the JIT (just-in-time) compiler for PTX files
+
+        library_globs = [
+            'libcuda.so.*',
+            'libnvidia-ptxjitcompiler.so.*',
+        ]
+        if LooseVersion(self.version) >= '11.5':
+            library_globs.append('libnvidia-nvvm.so.*')
+
+        startdir = self.cfg['start_dir']
+        nv_version = self.cfg['nv_version']
+        for library_glob in library_globs:
+            library_path = find_glob_pattern(os.path.join(startdir, library_glob))
+            library = os.path.basename(library_path)
+            # Sanity check the version
+            if library_glob == 'libcuda.so.*':
+                library_version = library.split('.', 2)[2]
+                if library_version != nv_version:
+                    raise EasyBuildError('Expected driver version %s (from nv_version) but found %s '
+                                         '(determined from file %s)', nv_version, library_version, library_path)
+
+            copy_file(library_path, os.path.join(libdir, library))
+            if library.endswith('.' + nv_version):
+                # E.g. libcuda.so.510.73.08 -> libcuda.so.1
+                versioned_symlink = library[:-len(nv_version)] + '1'
+            else:
+                # E.g. libnvidia-nvvm.so.4.0.0 -> libnvidia-nvvm.so.4
+                versioned_symlink = library.rsplit('.', 2)[0]
+            symlink(library, os.path.join(libdir, versioned_symlink), use_abspath_source=False)
+            # E.g. libcuda.so.1 -> libcuda.so
+            unversioned_symlink = versioned_symlink.rsplit('.', 1)[0]
+            symlink(versioned_symlink, os.path.join(libdir, unversioned_symlink), use_abspath_source=False)
+
+    def make_module_extra(self):
+        """Skip the changes from the Binary EasyBlock."""
+
+        return super(Binary, self).make_module_extra()
+
+    def sanity_check_step(self):
```

Comment:
You can use `nvidia-smi` as a sanity check here. That reports the supported CUDA version, and we should verify that we get what we expect here, for example:
```
nvidia-smi  -q --display=COMPUTE | grep CUDA | awk 'NF>1{print $NF}'
```

We should also check that the compat libraries actually work, as you must have a supported driver/compat combination. That's not trivial though.

Reply:
You could add CUDA as a build-dep, and compile the `devicequery` example and place it in the installation directory. That will error if you don't have a working combination.

Reply:
See the description: I'd rather do the test in the CUDA EC as this is easiest added as a dependency of CUDA. So you can have CUDA 11.3, 11.4, 11.5 & 11.6 depend on CUDAcompat 11.6. The other way round is harder: Which CUDA should be used as the build-dep? Not really good to install a CUDA version which is then not used or create cyclic dependencies.

However I'll add the sanity check. And in the test_step I'll add checking for a compatible driver. Both to be skipped if `nvidia-smi` isn't there (i.e. when the build machine is different than the deployment system)